### ISSUE TITLE
Error when running swift-algorithmsPackageTests.xctest: "unknown reloc type 16"

### ISSUE THREAD

**Issue by marcprux** [finagolfin/swift-android-sdk#175]
When building and running tests against `swift-algorithms` on an x86 host+emulator with the Swift 6.0 Android toolchain, I'm seeing the following warning and error:

```
WARNING: linker: /data/local/tmp/swift-android/swift-algorithms/swift-algorithmsPackageTests.xctest: unsupported flags DT_FLAGS_1=0x8000001
CANNOT LINK EXECUTABLE "/data/local/tmp/swift-android/swift-algorithms/swift-algorithmsPackageTests.xctest": unknown reloc type 16 @ 0x7e9b7e351e48 (16114)
Error: Error terminated(code: 134) running command: /Users/runner/Library/Android/sdk/platform-tools/adb shell /data/local/tmp/swift-android/swift-algorithms/swift-algorithmsPackageTests.xctest
```

This doesn't occur when running on an arm host+emulator (the tests all run and pass on my machine).

It's especially odd, since this package is passing tests on [CI](https://github.com/finagolfin/swift-android-sdk/actions).

Is this something that [termux-elf-cleaner](https://github.com/termux/termux-elf-cleaner) might help with? I notice that we are using it in CI, but only for the `swift-argument-parser` tests and not any other packages.

**Comment by finagolfin** [comment]
What Android APIs are you running it on in the emulator? List out what APIs you tried for each of arm and x86.

> Is this something that [termux-elf-cleaner](https://github.com/termux/termux-elf-cleaner) might help with?

No, I tried, it doesn't. This is related to the TLS issue I mentioned earlier.

**Comment by marcprux** [comment]
> This is related to the TLS issue I mentioned earlier.

FTR, the issue mentioned is: https://github.com/finagolfin/swift-android-sdk/pull/170#issuecomment-2351990840

**Comment by marcprux** [comment]
> What Android APIs are you running it on in the emulator? List out what APIs you tried for each of arm and x86.

I haven't tested against arm API variations yet, but on x86, testing swift-algorithms fails against API 25–28, and passes on API 29+ (see https://github.com/marcprux/swift-algorithms/actions/runs/11114132859). This tracks with the note about TLS becoming available in API 29: https://github.com/aosp-mirror/platform_bionic/blob/master/android-changes-for-ndk-developers.md#elf-tls-available-for-api-level--29

**Comment by finagolfin** [comment]
> I haven't tested against arm API variations yet

OK, but you tested it on _some_ AArch64 device or emulator, what API was that? I just wonder if it works before API 29 on AArch64.

> This tracks with the note about TLS becoming available in API 29

Yep, this is why [I raised the API used on the CI emulator from 24 to 29 with Swift 6](https://github.com/finagolfin/swift-android-sdk/commit/a0ebe225e979ce211e1d9207537282a89fe2ea11#diff-76992bc322b65ab94f299c05ba3cbd8b3ab2e32ebfdde6db5adba70fa9f5a26aR490), because all the test runners were failing otherwise.

The problem is that a Swift file in Foundation calls that [inline `_swift_stdlib_gettid()` function from the stdlib's synchronization shims that now uses a TLS variable](https://github.com/swiftlang/swift/pull/71383/files#diff-4a4beab0106e7043768f49f30baf61eb45116edabc020f16d39259ec2f9b6c6aR26), and that C++ code is presumably compiled by the Swift compiler somehow _without_ specifying an API level, which on Android then defaults to the latest API level, ie with native TLS.

We'll need to find where in the Swift compiler such inline compilation of C++ functions is done and check what it's doing wrong. The full target triple with the API level is passed to the `ClangImporter` that integrates C/C++ headers, so I'm not sure why this C++ compilation doesn't.

**Comment by marcprux** [comment]
> OK, but you tested it on _some_ AArch64 device or emulator, what API was that? I just wonder if it works before API 29 on AArch64.

It was API 34.

**Comment by marcprux** [comment]
I don't recall when this changed, but it works now. Closing.

**Comment by finagolfin** [comment]
[Doesn't appear to be the case](https://github.com/finagolfin/swift-android-sdk/actions/runs/11615927641/job/32347699382#step:38:134), where did you see it working?

**Comment by marcprux** [comment]
> [Doesn't appear to be the case](https://github.com/finagolfin/swift-android-sdk/actions/runs/11615927641/job/32347699382#step:38:134), where did you see it working?

Sorry, you are right. I was accidentally running against Android API 29 rather than 28 as I had thought.

**Comment by finagolfin** [comment]
I've submitted a fix for this upstream, swiftlang/swift#77883.

**Comment by marcprux** [comment]
Not sure whether this should be a separate issue report or not, but I was just testing [swift-http-types](https://github.com/apple/swift-http-types) with the pending 6.0.3 SDK against an API 34 aarch64 emulator and hit an error that sounds related to the TLS issue:

```
error: "./swift-http-typesPackageTests.xctest": executable's TLS segment is underaligned: alignment is 4, needs to be at least 64 for ARM64 Bionic
```

Not that it _doesn't_ occur when running against an API 29 x86_64 emulator with the 6.0.2 SDK on my [CI](https://github.com/marcprux/swift-http-types/actions/runs/12305486163/job/34345178181) for the project.

swift-http-types is a pretty simple package with no dependencies. I haven't tried to debug, but I suspect it might be something to do with [NIOLock.swift](https://github.com/apple/swift-http-types/blob/main/Sources/HTTPTypes/NIOLock.swift).

**Comment by finagolfin** [comment]
Yeah, that's related to this. Unfortunately, the linked fix will have to go into the compiler first before we can use it, will try to get that in for the next patch release.

**Comment by marcprux** [comment]
I expect you already know this, but FYI it is [this](https://github.com/apple/swift-http-types/blob/d57e30afe0c767c54b1213d96d17375a5c2edcfd/Sources/HTTPTypes/HTTPFields.swift#L112-L123) reference to `Synchronization.Mutex` that triggers the crash:

```
    #if compiler(>=6.0)
    @available(macOS 15.0, iOS 18.0, watchOS 11.0, tvOS 18.0, visionOS 2.0, *)
    private final class _StorageWithMutex: _Storage, @unchecked Sendable {
        let mutex = Mutex<Void>(())

        override func withLock<Result>(_ body: () throws -> Result) rethrows -> Result {
            try self.mutex.withLock { _ in
                try body()
            }
        }
    }
    #endif  // compiler(>=6.0)
```

Removing it (and fixing the code that references it to use the fallback NIOLock) lets the tests run (and pass) on my aarch64 emulator.

**Comment by finagolfin** [comment]
Got the emulated TLS fix for this into the release/6.0 branch upstream, will close this once it ships in the next patch release and we can run the release branch tests in the emulator at API 24 again, as I've already been able to do in trunk.

**Comment by finagolfin** [comment]
I just released a 6.1 bundle, which no longer has this problem and has been tested in the API 24 emulator for months now.
